File: 4 - Cuarto Semestre/4 - Estructuras de Datos y Algoritmos/4 - Evaluaciones/E1/bag.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Bag {
    private List<Object> items;

    public Bag() {
        items = new ArrayList<>();
        items.add(null);
        items.add(3);
        items.add(null);
        items.add(3);
        items.add(3);
        items.add(null);
    }

    public Object removeItem() throws Exception {
        for (int i = 0; i < items.size(); i++) {
            if (items.get(i) != null) {
                Object x = items.get(i);
                items.set(i, null);
                return x;
            }
        }
        throw new Exception("La bolsa esta vacia");
    }

    public Object removeRandom() throws Exception {
        Random rand = new Random();
        int pos = rand.nextInt(items.size());
        int n = (int)items.stream().filter(item -> item == null).count();
        if (n == items.size()) {
            throw new Exception("Bolsa vacia");
        }
        while (items.get(pos) == null) {
            pos++;
            if (pos == items.size()) {
                pos = 0;
            }
        }
        Object t = items.get(pos);
        items.set(pos, null);
        return t;
    }

    public int size() {
        return items.size() - (int)items.stream().filter(item -> item == null).count();
    }

    public void add(Object item) throws Exception {
        for (int i = 0; i < items.size(); i++) {
            if (items.get(i) == null) {
                items.set(i, item);
                return;
            }
        }
        throw new Exception("La bolsa esta llena");
    }

    public static void main(String[] args) throws Exception {
        Bag b = new Bag();

        int ant = b.size();
        b.add("Hola");
        int pos = b.size();
        assert(b.items.contains("Hola"));
        assert(pos - ant == 1);

        b.add("Hola");
        b.add("Hola");
        b.add("Hola");
    }
}

File: 4 - Cuarto Semestre/4 - Estructuras de Datos y Algoritmos/4 - Evaluaciones/E1/bag.py
import random

class Bag:

    def __init__(self):
        self.items = [ None, 3, None, 3, 3, None ]

    def removeItem(self):
        for i in range(len(self.items)):
            if self.items[i] is not None:
                x = self.items[i]
                self.items[i] = None
                return x
        raise Exception("La bolsa esta vacia")

    def removeRandom(self):
        pos = random.randint(0,len(self.items))
        n = self.items.count(None)
        if n==len(self.items):
            raise Exception("Bolsa vacia")
        while self.items[pos]==None:
            pos += 1
            if pos==len(self.items):
                pos = 0
        t = self.items[pos]
        self.items[pos]=None
        return t

    def size(self):
        return len(self.items)-self.items.count(None)
    
    def add(self, item):
        for i in range(len( self.items)):
            if self.items[i] is None:
                self.items[i] = item
                return
        raise Exception("La bolsa esta llena")


if __name__ == "__main__":
    b = Bag()

    ant = b.size()
    b.add("Hola")
    pos = b.size()
    assert( "Hola" in b.items )
    assert( pos-ant == 1 )

    b.add("Hola")
    b.add("Hola")
    b.add("Hola")